[E0284] suggestion is useless

### Code

```Rust
trait S {
    type Assoc<U>: T;
    fn f<B, F>(self, f: Self::Assoc<F>) -> Self::Assoc<B>;
}

trait T: S { }

fn g<A>(a: A) where A: T {
    a.f::<A::Assoc<A>, _>(todo!());
}
```

### Current output

```Shell
error[E0282]: type annotations needed
 --> src/lib.rs:9:7
  |
9 |     a.f::<A::Assoc<A>, _>(todo!());
  |       ^ cannot infer type of the type parameter `F` declared on the method `f`
  |
help: consider specifying the generic arguments
  |
9 -     a.f::<A::Assoc<A>, _>(todo!());
9 +     a.f::<<A as S>::Assoc<A>, _>(todo!());
  |
```

### Desired output

```Shell

```

### Rationale and extra context

The suggestion is just rewriting it and won't change anything.

### Other cases

```Rust

```

### Rust Version

```Shell
rustc 1.94.0 (4a4ef493e 2026-03-02)
binary: rustc
commit-hash: 4a4ef493e3a1488c6e321570238084b38948f6db
commit-date: 2026-03-02
host: x86_64-unknown-linux-gnu
release: 1.94.0
LLVM version: 21.1.8
```

### Anything else?

_No response_

--- a/compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
+++ b/compiler/rustc_trait_selection/src/error_reporting/infer/need_type_info.rs
@@ -1030,12 +1030,15 @@ impl<'a, 'tcx> FindInferSourceVisitor<'a, 'tcx> {
                         let args = self.node_args_opt(expr.hir_id)?;
                         let span = tcx.hir_span(segment.hir_id);
                         let insert_span = segment.ident.span.shrink_to_hi().with_hi(span.hi());
+                        let have_turbofish = segment.args.is_some_and(|args| {
+                            args.args.iter().any(|arg| arg.is_ty_or_const())
+                        });
                         InsertableGenericArgs {
                             insert_span,
                             args,
                             generics_def_id: def_id,
                             def_id,
-                            have_turbofish: false,
+                            have_turbofish,
                         }
                     };
                     return Box::new(insertable.into_iter());

--- a/tests/ui/inference/useless-turbofish-suggestion.rs
+++ b/tests/ui/inference/useless-turbofish-suggestion.rs
@@ -1,0 +1,28 @@
+// Regression test for #153732.
+//
+// When a method call already has turbofish type arguments, don't suggest
+// rewriting them — the suggestion just rewrites user syntax into
+// fully-qualified form without resolving anything.
+//
+// The span still points at the method name rather than the unresolved `_`;
+// fixing that is left as future work.
+
+struct S;
+
+impl S {
+    fn f<A, B>(self, _a: A) -> B {
+        todo!()
+    }
+}
+
+fn with_turbofish() {
+    S.f::<u32, _>(42);
+    //~^ ERROR type annotations needed
+}
+
+fn without_turbofish() {
+    S.f(42);
+    //~^ ERROR type annotations needed
+}
+
+fn main() {}

--- a/tests/ui/inference/useless-turbofish-suggestion.stderr
+++ b/tests/ui/inference/useless-turbofish-suggestion.stderr
@@ -1,0 +1,20 @@
+error[E0282]: type annotations needed
+  --> $DIR/useless-turbofish-suggestion.rs:19:7
+   |
+LL |     S.f::<u32, _>(42);
+   |       ^ cannot infer type of the type parameter `B` declared on the method `f`
+
+error[E0282]: type annotations needed
+  --> $DIR/useless-turbofish-suggestion.rs:24:7
+   |
+LL |     S.f(42);
+   |       ^ cannot infer type of the type parameter `B` declared on the method `f`
+   |
+help: consider specifying the generic arguments
+   |
+LL |     S.f::<i32, B>(42);
+   |        ++++++++++
+
+error: aborting due to 2 previous errors
+
+For more information about this error, try `rustc --explain E0282`.

--- a/tests/ui/issues/issue-23041.stderr
+++ b/tests/ui/issues/issue-23041.stderr
@@ -3,12 +3,6 @@ error[E0282]: type annotations needed
    |
 LL |     b.downcast_ref::<fn(_)->_>();
    |       ^^^^^^^^^^^^ cannot infer type of the type parameter `T` declared on the method `downcast_ref`
-   |
-help: consider specifying the generic argument
-   |
-LL -     b.downcast_ref::<fn(_)->_>();
-LL +     b.downcast_ref::<fn(_) -> _>();
-   |
 
 error: aborting due to 1 previous error
 

